src/opensles/AudioInputStreamOpenSLES.cpp

```diff
@@ -0,0 +1,219 @@
+            (SLuint32) chanCountToChanMask(mChannelCount), // channelMask
+            getDefaultByteOrder(),
+    };
+
+    SLDataSink audioSink = {&loc_bufq, &format_pcm};
+
+    /**
+     * API 21 (Lollipop) introduced support for floating-point data representation and an extended
+     * data format type: SLAndroidDataFormat_PCM_EX. If running on API 21+ use this newer format
+     * type, creating it from our original format.
+     */
+    SLAndroidDataFormat_PCM_EX format_pcm_ex;
+    if (__ANDROID_API__ >= __ANDROID_API_L__) {
+        SLuint32 representation = OpenSLES_ConvertFormatToRepresentation(getFormat());
+        // Fill in the format structure.
+        format_pcm_ex = OpenSLES_createExtendedFormat(format_pcm, representation);
+        // Use in place of the previous format.
+        audioSink.pFormat = &format_pcm_ex;
+    }
+
+    // configure audio source
+    SLDataLocator_IODevice loc_dev = {SL_DATALOCATOR_IODEVICE,
+                                      SL_IODEVICE_AUDIOINPUT,
+                                      SL_DEFAULTDEVICEID_AUDIOINPUT,
+                                      NULL };
+    SLDataSource audioSrc = {&loc_dev, NULL };
+
+    SLresult result = EngineOpenSLES::getInstance()->createAudioRecorder(&mObjectInterface,
+                                                                       &audioSrc,
+                                                                       &audioSink);
+    if (SL_RESULT_SUCCESS != result) {
+        LOGE("createAudioRecorder() result:%s", getSLErrStr(result));
+        goto error;
+    }
+
+    // Configure the voice recognition preset, which has no
+    // signal processing, for lower latency.
+    SLAndroidConfigurationItf inputConfig;
+    result = (*mObjectInterface)->GetInterface(mObjectInterface,
+                                            SL_IID_ANDROIDCONFIGURATION,
+                                            &inputConfig);
+    if (SL_RESULT_SUCCESS == result) {
+        SLuint32 presetValue = SL_ANDROID_RECORDING_PRESET_VOICE_RECOGNITION;
+        (*inputConfig)->SetConfiguration(inputConfig,
+                                         SL_ANDROID_KEY_RECORDING_PRESET,
+                                         &presetValue,
+                                         sizeof(SLuint32));
+    }
+
+    result = (*mObjectInterface)->Realize(mObjectInterface, SL_BOOLEAN_FALSE);
+    if (SL_RESULT_SUCCESS != result) {
+        LOGE("Realize recorder object result:%s", getSLErrStr(result));
+        goto error;
+    }
+
+    result = (*mObjectInterface)->GetInterface(mObjectInterface, SL_IID_RECORD, &mRecordInterface);
+    if (SL_RESULT_SUCCESS != result) {
+        LOGE("get recorder interface result:%s", getSLErrStr(result));
+        goto error;
+    }
+
+    result = AudioStreamOpenSLES::registerBufferQueueCallback();
+    if (SL_RESULT_SUCCESS != result) {
+        LOGE("get bufferqueue interface:%p result:%s", mSimpleBufferQueueInterface, getSLErrStr(result));
```

Comment:
Why isn't this error message being printed by registerBufferQueueCallback? You seem to have the same code in the output stream.

Reply:
It is printed in registerBufferQueueCallback() so I removed it from the input and output streams.